src/mbgl/map/map.cpp

```diff
@@ -276,46 +249,70 @@ void Map::Impl::update() {
 
     style->updateTiles(parameters);
 
-    if (mode == MapMode::Continuous) {
-        backend.invalidate();
-    } else if (stillImageRequest && style->isLoaded()) {
-        // TODO: determine whether we need activate/deactivate
-        BackendScope guard(backend);
-        render(stillImageRequest->view);
-    }
-
     updateFlags = Update::Nothing;
-}
 
-void Map::Impl::render(View& view) {
     if (!painter) {
         painter = std::make_unique<Painter>(backend.getContext(), transform.getState(), pixelRatio);
     }
 
-    FrameData frameData { timePoint,
-                          pixelRatio,
-                          mode,
-                          contextMode,
-                          debugOptions };
+    if (mode == MapMode::Continuous) {
+        if (renderState == RenderState::Never) {
+            backend.notifyMapChange(MapChangeWillStartRenderingMap);
+        }
 
-    painter->render(*style,
-                    frameData,
-                    view,
-                    annotationManager->getSpriteAtlas());
+        backend.notifyMapChange(MapChangeWillStartRenderingFrame);
+
+        FrameData frameData { timePoint,
+                              pixelRatio,
+                              mode,
+                              contextMode,
+                              debugOptions };
+
+        painter->render(*style,
+                        frameData,
+                        view,
+                        annotationManager->getSpriteAtlas());
+
+        painter->cleanup();
```

Comment:
In the static rendering case, we're calling `cleanup` after the callback (meaning that we don't have to wait for the cleanup to happen before the invoking implementation can do something with the result)